packages/paymasters/contracts/PermitERC20UniswapV3Paymaster.sol

```diff
@@ -147,33 +239,116 @@ contract PermitERC20UniswapV3Paymaster is BasePaymaster, ERC2771Recipient {
     internal
     override
     {
-        (address payer,) = abi.decode(context, (address, uint256));
-        uint256 ethActualCharge = relayHub.calculateCharge(gasUseWithoutPost + gasUsedByPost, relayData);
-        _depositProceedsToHub(ethActualCharge);
-        uint256 remainingTokenBalance = token.balanceOf(address(this));
-        require(token.transfer(payer, remainingTokenBalance), "failed refund");
-        emit TokensCharged(gasUseWithoutPost, gasUsedByPost, ethActualCharge);
-    }
-
-    function _depositProceedsToHub(uint256 ethActualCharge) private {
-        UniswapV3Helper.swapToEth(
-            address(weth),
-            address(token),
-            ethActualCharge,
-            uniswapPoolFee,
-            uniswap
-        );
-        relayHub.depositFor{value : address(this).balance}(address(this));
+        (address payer, uint256 priceQuote, uint256 tokenPreCharge) = abi.decode(context, (address, uint256, uint256));
+        IERC20Metadata token = _getTokenFromPaymasterData(relayData.paymasterData);
+        (uint256 tokenActualCharge, uint256 ethActualCharge) = _calculateCharge(relayData, gasUseWithoutPost + gasUsedByPost, priceQuote, tokensSwapData[token].priceDivisor);
```

Comment:
we assume that gasUsedByPost is constant, regardless of token in use.
I would like to keep it this way (to avoid the complexity of calculating it separately for each token), but we do need to verify that at least for the known tokens, it is close enough (that is, the total cost of of the relayed TX doesn't fluctuate too much with different tokens.)